src/frontends/lean/elaborator.h

```diff
@@ -246,23 +246,113 @@ class elaborator {
+        /** Name of the structure that is the source of the field. Is ancestor of \c m_struct_name */
+        name m_base_struct_name;
+        /** Name of the structure for the projected expression */
+        name m_struct_name;
+        /** The field name for the projection */
+        name m_field_name;
+
+        field_resolution_proj_fn(name const & base_struct_name, name const & struct_name, name const & field_name):
+            m_base_struct_name(base_struct_name), m_struct_name(struct_name), m_field_name(field_name) {}
+
+        /** Get the name of the projection function */
+        name get_full_name() const { return m_base_struct_name + m_field_name; }
+    };
+
+    /** Field resolution: this is a "method" (a.k.a. extended dot notation) */
+    struct field_resolution_const {
+        /** If this is not equal to \c m_struct_name then we should insert parent projections, and this is
+         * the name of the structure that is the source of the field. Is ancestor of \c m_struct_name */
+        name m_base_struct_name;
+        /** Generalized structure name for the method expression */
+        name m_struct_name;
+        /** Name of the constant to use as a function */
+        name m_const_name;
+
+        field_resolution_const(name const & base_struct_name, name const & struct_name, name const & const_name):
+            m_base_struct_name(base_struct_name), m_struct_name(struct_name), m_const_name(const_name) {}
+    };
+
+    /** Field resolution: projection is being used to make a "local" recursive call */
+    struct field_resolution_local_rec {
+        /** The generalized structure name for the argument to the call. */
+        name m_base_name;
+        /** The resolved name for the recursive call (for error reporting). */
+        name m_full_name;
+        /** The declaration for the "local" recursive call. */
+        local_decl m_ldecl;
+
+        field_resolution_local_rec(name const & base_name, name const & full_name, local_decl const & ldecl):
+            m_base_name(base_name), m_full_name(full_name), m_ldecl(ldecl) {}
+    };
+
     struct field_resolution {
-        name m_S_name; // structure name of field expression type
-        name m_base_S_name; // structure name of field
-        name m_fname;
-        optional<local_decl> m_ldecl; // projection is a local constant: recursive call
-
-        field_resolution(name const & full_fname, optional<local_decl> ldecl = {}):
-                m_S_name(full_fname.get_prefix()), m_base_S_name(full_fname.get_prefix()),
-                m_fname(full_fname.get_string()), m_ldecl(ldecl) {}
-        field_resolution(const name & S_name, const name & base_S_name, const name & fname):
-                m_S_name(S_name), m_base_S_name(base_S_name), m_fname(fname) {}
-
-        name get_full_fname() const { return m_base_S_name + m_fname; }
+        enum class kind { ProjFn, Const, LocalRec };
+
+        kind m_kind;
+        union {
```

Comment:
I'd rather avoid unions.  Could we just inline all the fields instead?

Reply:
Ok, I've inlined them. I kept the structures for the three types, though, since they're a nicer interface than trying to use constructors, and they seem like they might be better than using factory methods since they can hold onto documentation about fields.

Regarding mathlib, I've compiled part of it, and the only change so far has been the different handling of un-applied projections (where before they'd keep their implicit arguments). This detail isn't used often and it's easy to fix. I still need to compile all of mathlib -- the big changes to notation handling have made compiling it more difficult, but I'm going to try again with that change removed.

Reply:
So far a difference that's shown up is that if `hab : a < b` with `a b : ℤ` then you need to do `int.lt.dest hab` rather than `hab.dest` (which gives "error: invalid field notation, function 'int.lt.dest' does not have explicit argument with type (int.lt ...)"). This appears to match Lean 4 behavior using this test:
```lean
def Nat.lt.foo {a b : Nat} (h : a < b) := a

variable (h : 1 < 2)
#check h.foo
/-
invalid field notation, function 'Nat.lt.foo' does not have argument with type (Nat.lt ...) that can be used, it must be explicit or implicit with an unique name
-/
```

Reply:
It turns out this `h.foo` feature depends on a (mis?)-feature of dot notation, which is that when you have `h.foo` in a non-application position (like here) then it will insert `h` as the first argument. With the change to unify the behavior of applications and non-applications, this no longer works. It turns out this feature is used often in mathlib (for example, `a.of_dual` [here](https://github.com/leanprover-community/mathlib/blob/f5afe205c23d7e9f065a04390cc337dfdb7e6e65/src/order/bounded_order.lean#L895) manages to invoke the `has_coe_to_fun` instance for `with_top.of_dual : with_top αᵒᵈ ≃ with_bot α`).

For now, I've added a generalized version of this feature, where if the resolution procedure fails to find an explicit argument of the right type, it inserts it as the first argument. It's not compatible with Lean 4, though. We could artificially limit this feature to only apply when there are no arguments, matching the previous behavior, but it seems useful for implementing extension methods, which this and the alias method feature would give you. (Another design would be for dot notation to try synthesizing `has_coe_to_fun` instances when the resolved method is not a function.)